### ISSUE TITLE
💡 [Feature]: Extra tasks to update Patch, Minor and Major version of SPFx project

### ISSUE THREAD

**Issue by martinlingstuyl** [pnp/vscode-viva#94]
## 🎯 Aim of the feature

Add extra tasks so that you can update the patch, minor or major version of an SPFx project (package-solution.json) in one click. 
Extra nice would be if it would also update the package.json version as well.

## 📷 Images (if possible) with expected result

![image](https://github.com/pnp/vscode-viva/assets/5267487/ca2eb9ee-490c-4a45-b779-2a49977728b3)

**Comment by Adam-it** [comment]
Thanks @martinlingstuyl for adding an idea 👍.
A couple of things I would like to clarify:

- why do you usually update the version? Is it something you do just before the package publish to update it on the app catalog? Or is it something you just do regularly on adding every new feature and you just want to add this to the commit?
- do you always have aligned the package.json and package-solution.json? And if not what you would expect to happen on this task? Should one be overwritten to align with pacakge-solution.json? Or both should be increased?
- what number should be increased? Major? Minor? Or patch? Or should the extension ask for this option?

**Comment by martinlingstuyl** [comment]
Do my answers clarify it a bit for you? I'd wish to be able to be more specific on the impact of version numbers. maybe we should ask Vesa :-)

**Comment by Adam-it** [comment]
> Do my answers clarify it a bit for you? I'd wish to be able to be more specific on the impact of version numbers. maybe we should ask Vesa :-)

Yes that helped a lot. Thanks for the additional info. 
@VesaJuvonen is usually very busy (totally understandable), but if you have anything worth adding regarding versioning of apfx package please do 👍

**Comment by Adam-it** [comment]
@martinlingstuyl I will set this feature for v3 
In your answer I noticed 

>We use a gulp task to sync both numbers, so they are the same.

So you use a custom defined gulp task for it, right? I think in there is no OOTB gupl task for it delivered with @microsoft/sp-build-core-tasks package 🤔? Viva Toolkit base on those tasks as then we assume in every project they exist. We don't deliver any custom gulp tasks for every project.

**Comment by martinlingstuyl** [comment]
> I think in there is no OOTB gupl task for it delivered with @microsoft/sp-build-core-tasks package

No that's correct, we've copied it from internet somewhere. It's picking up on an npm event I believe.

**Comment by Adam-it** [comment]
> > I think in there is no OOTB gupl task for it delivered with @microsoft/sp-build-core-tasks package
> 
> No that's correct, we've copied it from internet somewhere. It's picking up on an npm event I believe.

hmmm.. not possible path for this extension. but we may for sure prototype a different approach

**Comment by Adam-it** [comment]
found an awesome inspiration coming from the good old pnp spfx generator 
https://n8d.at/use-npm-version-to-upgrade-version-of-your-spfx-solution/

- The idea I have is to create similar functionality with a dedicated action like 'increase version'. 
- When clicked it would present 3 options which version we want to increase, `major`, `minor`, `patch`
- It would always align the versions between `package.json` and `package-solution.json` without asking
- we could add a dedicated setting in the VS Code settings for SPFx toolkit 'update version when deploying solution'. This should be true/false check box. When set to true during the deploy action SPFx toolkit should now also ask which version should be updated: `major`, `minor`, `patch` and should perform this change before running deploy.

things to consider:
- should we also run `git tag` command to create a dedicated tag for the version update changes commit 🤔. That is what pnp spfx generator did
- our scaffolded CI/CD workflows for GitHub and AzureDevOps do not update this version on deploy. Maybe this should be also an option 🤔

@martinlingstuyl any feed? how do you like the above prototype 👆

**Comment by martinlingstuyl** [comment]
Hi @Adam-it,

I certainly love the idea! 

In terms of git tagging the commit: absolutely! Or maybe use a setting to toggle this on/off. Default should be true though, I think.

As far as the CI/CD workflow is concerned: I don't think this applies there. The CI/CD workflows simply pick the last commit, and build and deploy based on that. Which is what we would want.

**Comment by Adam-it** [comment]
ok sounds like a plan. 

to sum up the idea: 

1. Create a dedicated action 'increase version'.
2. When clicked it would present 3 options which version we want to increase, major, minor, patch
3. It would always align the versions between package.json and package-solution.json without asking
4. we could add a dedicated setting in the VS Code settings for SPFx toolkit 'update version when deploying solution'. This should be true/false check box. When set to true during the deploy action SPFx toolkit should now also ask which version should be updated: major, minor, patch and should perform this change before running deploy.

let's drop the idea for `git tag` for now and we may always get back to this idea later

**Comment by nicodecleyre** [comment]
Great feature! Can I work on this one?

**Comment by Adam-it** [comment]
> Great feature! Can I work on this one?

Sure thing. Please do remember to start and target only `dev` branch in your PRs. The `main` branch is only to track minor and major releases and `dev` branch is used for bet/pre-release and current development 👍

**Comment by nicodecleyre** [comment]
I have a question, when doing point 4 deploying the solutions and increase the version, it should bundle & package the solution before deploying the package to your app catalog, otherwise the increased number wont be shipped.

As we don't know if it should be bundled & pacakged for local or production, maybe we should prompt for it during deploy (and only if the version is increased)?

**Comment by Adam-it** [comment]
> I have a question, when doing point 4 deploying the solution and increase the version, it should bundle & package the solution before deploying the package to your app catalog, otherwise the increased number wont be shipped.
> 
> As we don't know if it should be bundled & packaged for local or production, maybe we should prompt for it during deploy (and only if the version is increased)?

hmm... good point. I forgot about that 🤦‍♂️
The problem with this is that it will now significantly extend the deploy action with the time that is needed to perform bundle and package. 
I wouldn't like to always bundle and package every time the deploy is clicked. This just does not make sense. I guess devs know what they do and should bundle and package the solution using the correct gulp task and then use the deploy action. 
hmmm.. this seems tricky 🤔
@nicodecleyre do you have any suggestions how we should proceed?
TBH I would NOT develop point 4 and maybe think of some alternatives. 
The worst case would be that the dev will need to do the following operations (each step is a separate click):
1. use the increase number action
2. click the correct gulp bundle
3. click the correct gulp package
4. click the delpoy

what we could do in future (not part of this issue but maybe as a follow up) is to perform an another action that would combine all of those 4 steps into single click. In the past the PnP Yo generator had a task like `publish` that combined `bundle` & `package` into single task 🤔. Maybe we could do something similar but in our case `publish` would ask which version number should be incremented and then which type of bundle you want to do `local` or `production` and then it would do all of those 4 steps 🤔. What is your take on that? 

@nicodecleyre so to sum up. Lets drop point 4 for now and not implement it with this issue. (I crossed it out in my previous post with the plan) and think of an alternative functionality. For example something what I suggested above.

**Comment by nicodecleyre** [comment]
Fair point that it would extend the deploy action significantly. Maybe we should move point 4 to the bundle action? As you would always have to do bundle, then package-solution, then deploy when the version number is increased. So that way it would prompt to change the version number when bundle  the solution and when the setting is checked to ask to increase the version number.

Yes, definitely, I was thinking the same, there should be a publish task which would combine bundle, package and perhaps clean (at least i'm doing a gulp clean everytime before starting with bundle 😝 )  - together with that i was also thinking the we maybe could remove build local, build production, package solution local, package solution production and replace them with just 2 buttons "build" & "package", which can ask you if you want to build/package for local or for production environment (but this is not part of this issue ofc 😄 )

**Comment by Adam-it** [comment]
The related PR got just merged. This will be available in a pre-release quite soon. 
Thank you for your amazing work 👏
You Rock 🤩
